### ISSUE TITLE
Specifying desired version of WebDriver binary

### ISSUE THREAD

**Issue by jugglinmike** [web-platform-tests/wpt#12361]
Currently, `wpt run {product}` and `wpt install {product} webdriver` install
the latest version of the WebDriver server for the specified `product`. This
seems like good default behavior, but it is insufficient for two use cases:

- Reproducing results - whether verifying data published online or re-trying a
  failed trial
- Distributing execution over time - the CLI's "chunking" feature allows for
  the test suite to be run in segments, possibly over many hours. New WebDriver
  releases published during this interval could cause the final merged result
  set to reflect a "blend" of multiple binaries

Matching WebDriver releases with the browser releases they support is another
problem, but we may be able to address that separately. I think we can support
those use cases with zero impact on existing workflows via an extension to the
`wpt install` command:

    $ ./wpt run {product}
    # No change to current behavior:
    #   if no binary is present, install the latest version

    $ ./wpt install {product} webdriver
    # No change to current behavior:
    #   install the latest version

    $ ./wpt install {product} webdriver {version}
    # New feature:
    #   install the specified version

@jgraham @gsnedders what do you think?

**Comment by gsnedders** [comment]
I'm not opposed, though I'm dubious how useful it is (i.e., I think the benefit is marginal).

**Comment by jgraham** [comment]
I think all of these concerns apply equally to the browser itself, but are more relvant there (it seems much more likely that there is a new browser nightly during the interval between starting tests than that there is a new webdriver frontend).

For TC at least I would probably attack this by having a single task that downloaded the binaries to artifacts, and then have the other tasks grab those artifacts (or maybe just have it grab the URLs and pass those on to other tasks). This would ensure that the remaining tasks were all synchronised on version, and that tasks could be rerun later with the same version.

That obviously doesn't cover other chunked running scenarios, but it might be good enough.

**Comment by gsnedders** [comment]
And we have even bigger reproducibility problems with Safari and pre-release Edge builds (given AFAIK at least for Edge you can in theory manually install all the patches from the base Win10 install to get an equivalent setup).

**Comment by jugglinmike** [comment]
> I think all of these concerns apply equally to the browser itself, but are more relvant there (it seems much more likely that there is a new browser nightly during the interval between starting tests than that there is a new webdriver frontend).

Agreed, though Chrome/Chromium doesn't commit to a stable URL for experimental builds, so we can't offer this functionality for all browsers. That's why I'm limiting this to WebDriver binaries.

> For TC at least I would probably attack this by [...]

[That's exactly what we're doing in the BuildBot system today](https://github.com/web-platform-tests/results-collection/blob/a2d6731ffa1cb69bb584a479425ae0f30846b3ff/src/scripts/get-browser-url.py) :) Come to think of it, I could just implement a corresponding script for the WebDriver binaries. Maybe those scripts will find their way into WPT's TaskCluster integration one day.

> And we have even bigger reproducibility problems with Safari and pre-release Edge builds

Yup, that's also why we don't implement `wpt install edge`. OS-bundled browsers are out of scope for the WPT CLI, but I don't think that should limit the CLI's utility for other browsers.
